### ISSUE TITLE
Track changes on cloned streams

### ISSUE THREAD

**Issue by martinthomson** [w3c/mediacapture-main#271]
@pehrsons notes [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1208371#c210) an issue with stream cloning.

The fundamental question raised is whether changes to the set of tracks on the original stream should be propagated to clones of that stream.  The [current algorithm](https://w3c.github.io/mediacapture-main/#widl-MediaStream-clone-MediaStream) does not propagate changes.

It's very tempting to suggest that the current algorithm is sufficient and that changes can be managed by the application ([I did](https://bugzilla.mozilla.org/show_bug.cgi?id=1208371#c209)), but @pehrsons raises a few points that I think need more consideration.  These concerns arise from the fact that the set of tracks on a stream can change automatically, whether as a result of track additions by `RTCPeerConnection` or maybe `HTMLMediaElement::captureStream()`.

The simplest polyfill looks like this:

```js
var cloneStream = originalStream.clone();
originalStream.onaddtrack = e => cloneStream.addTrack(e.track.clone());
originalStream.onremovetrack = e => cloneStream.removeTrack(/* something more complex */);
```

In cases where the browser modifies the set of tracks, use of the clone and use of the original will differ in ways that cannot be easily repaired by an application.  If a new track is added, the time required to propagate the `onaddtrack` event means that an application-managed clone could miss the lead-in for a track on the clone.  A similar concern applies to `onremovetrack`.

**Comment by jan-ivar** [comment]
> `originalStream.onremovetrack = e => cloneStream.removeTrack(/* something complex */);`

Removing is easy, since presumably tracks disappear everywhere when a host pulls them, meaning `onremovetrack` fires on clones and originals alike, so this line wouldn't be needed.

So we can narrow the issue to adding of new tracks and `onaddtrack` I believe. While other specs may be able to work around us, it would probably help if we thought through how we'd expect this to work, and gave guidance, so there'd be consistency.

**Comment by martinthomson** [comment]
Well, if the tracks merely end, this shouldn't matter because ending will affect all clones, but removal for other reasons isn't as straightforward.

**Comment by alvestrand** [comment]
1) Huh?
I thought we decided long ago that when you clone a stream, you get a new stream. Nothing that happens to the old stream affects the new one, except indirectly through the tracks they share.

2) I haven't seen anything that says tracks that end are removed from the stream. If a track in a stream ends, you end up with a stream having an ended track.

What am I missing?

**Comment by jan-ivar** [comment]
@alvestrand I think you're missing [this line](http://w3c.github.io/mediacapture-main/getusermedia.html#mediastreamtrackevent):

"The addtrack and removetrack events notify the script that the track set of a MediaStream has been updated by the User Agent."

AFAICT this gives browsers the ability to add and remove tracks to any stream at any time for any reason. This impedance mismatch exists to support host-managed streams in other specs. Other specs need to solve what happens to streams and stream-clones in these cases, and need guidance. For instance:

In [WebRTC](http://w3c.github.io/webrtc-pc/#events-on-mediastream) what should happen to stream-clones when I do this mid-call:

    pc.removeTrack(stream.getTracks()[0]); pc.addTrack(someNewTrack, stream); // +renegotiate

In [mediacapture-fromelement](https://w3c.github.io/mediacapture-fromelement/index.html), common things like looping and source manipulation cause tracks to get swapped out of the stream.

**Comment by jan-ivar** [comment]
Nevermind that removeTrack takes a sender, but you know what I mean.

**Comment by stefhak** [comment]
I think the mediacapture-fromelement document specifies tracks coming and going out pretty well, there will be removeTrack and addTrack events fired on the MediaStream, and a media element videoTrack or audioTrack that goes a way and then comes back will be (when coming back) be represented by a new MediaStreamTrack (as ended MSTs can't come back to life).

As for the PeerConnection example, the sending side app controls what will happen on the remote side via the streams argument to addTrack. The only uncertain thing I can see in your example is the order the addTrack and removeTrack events should fire on the remote MediaStream.

So that deals with @alvestrand #2 above. For @alvestrand #1 I think Harald is correct.

**Comment by alvestrand** [comment]
@jan-ivar my response bullet 2 about ending was in response to @martinthomson's statement "Well, if the tracks merely end, this shouldn't matter because ending will affect all clones, but removal for other reasons isn't as straightforward." - I read this as claiming that ending was a reason for removal, which didn't make sense to me. Mediacapture-fromelement notes that MediaStreamTracks that are removed should also end; that's not quite the same thing.
We may be all right here.

**Comment by martinthomson** [comment]
I don't think that the original problem is getting through here and I don't know how I can rephrase this to help.  Certainly, my "removal for other reasons" comment only muddied the waters.  I'll try again anyway...

If you remove a track from a stream, then it stops being present in the playback of that stream*.  If you add a track to a stream, then it starts being present in the playback of that stream*. [*] modulo track selection and that sort of thing.

It is entirely possible that the events that these generate happen after the effects can be observed.  For example, if the browser adds a track to a stream, you might see the contents of that track before the `addtrack` event fires.

For a cloned stream, if we rely on an application propagating track changes between cloned streams, changes of this nature will result in a delay between the manifestation of the change in the original stream and the manifestation of the change in the cloned stream.

For example, I take a stream from `RTCPeerConnection` and clone it.  It initially contains only video.  A live audio track is added.  The original stream plays out the phrase "Hello, my name is Lucy.".  The application has to propagate the changes before the cloned stream sees the track and by the time this is done it plays out only "I name is Lucy".

**Comment by stefhak** [comment]
We moved away from that problem a bit, but it was not forgotten!
I'm trying to find the use case where this is a practical problem. I don't think the `RTCPeerConnection` example is one. First of all, the time for the signaling to set up the additional audio track is probably much longer than the event propagation on the receiving side. Secondly, we have ways to set up an empty track from the start, and make that part of both streams on the remote side (and once content starts flowing in that track it would do so in both streams of course).
`HTMLMediaElement::captureStream()`could be a case, but why can't the original stream be used for all consumers (instead of being cloned)? Or why can't the second consumer capture from the same element rather than cloning the first stream being captured?

**Comment by martinthomson** [comment]
>  I'm trying to find the use case where this is a practical problem. I don't think the RTCPeerConnection example is one. First of all, the time for the signaling to set up the additional audio track is probably much longer than the event propagation on the receiving side.

Aren't there cases where streams appear prior to receiving signaling?  I'm thinking about the case of a re-offer with bundle.  The answerer can start sending media long before passing back an answer.

> Secondly, we have ways to set up an empty track from the start, and make that part of both streams on the remote side (and once content starts flowing in that track it would do so in both streams of course).

That's part of the solution, yes.

> HTMLMediaElement::captureStream()could be a case, but why can't the original stream be used for all consumers (instead of being cloned)? Or why can't the second consumer capture from the same element rather than cloning the first stream being captured?

Well, that's questioning the need for `MediaStream::clone()`.  I'm starting to be convinced by your arguments here.  If we can be satisfied that `onaddtrack` propagation is not needed, then I think we could say the same for `clone()`.

**Comment by jan-ivar** [comment]
I think I agree as well, mostly because I can't come up with how it would work otherwise. The only reason to clone a stream in the first place is wanting control over what's in and what's out, which I can't reconcile with automatic track addition (If I clone a DOMElement-captured stream to remove, say, the audio track, do I get new audio tracks or not when the source loops?).

There really are two kinds of streams here: hosted vs plain, and it sounds like we're agreeing that cloning a hosted stream should give you a plain stream, not a hosted stream.

We should remove `stream.clone()` then, to avoid confusion.

`var clone = new MediaStream(stream);` should suffice.

**Comment by martinthomson** [comment]
@jan-ivar, I think that your point about cloning is largely right.  Unless you want to apply new constraints on the cloned tracks. But then it we propagated track changes to clones, you would have to deal with having an unconstrained track for some time, which might not be compatible with that model.

`var clone = new MediaStream(stream);` doesn't quite do the same, unless we state that the tracks are cloned by the constructor.  It currently does not.

**Comment by jan-ivar** [comment]
That teaches me to read the spec. I guess `stream.clone()` still has value then. I don't think constraints work on hosted tracks, do they?

**Comment by martinthomson** [comment]
I don't know what you mean by a hosted track.  The constraints API points are visible on all `MediaStreamTrack` instances, but applying constraints might not work if the source doesn't support the constraint.  A lot of the constraints should work everywhere.  For instance, I think that you can always downscale video or downsample audio.

**Comment by jan-ivar** [comment]
I see no support for the generalization that a lot of constraints should work everywhere. Constraints do not work in the cases discussed here. [e.g.](http://w3c.github.io/mediacapture-main/getusermedia.html#the-model-sources-sinks-constraints-and-settings): "sources with the remote attribute set to true do not consider the current constraints applied to a track".

There is no mention at all of constraints in https://w3c.github.io/mediacapture-fromelement

Firefox does not downscale camera resolutions (though Chrome does), which means that except on OSX which unavoidably downscales and upscales camera resolutions, the camera resolutions returned by Firefox are guaranteed to be native (to the driver at least). We have that fitness distance model remember?

**Comment by stefhak** [comment]
>>  I'm trying to find the use case where this is a practical problem. I don't think the RTCPeerConnection example is one. First of all, the time for the signaling to set up the additional audio track is probably much longer than the event propagation on the receiving side.

> Aren't there cases where streams appear prior to receiving signaling?  I'm thinking about the case of a re-offer with bundle.  The answerer can start sending media long before passing back an answer.

I've been thinking a bit about that. (If the answerer stars sending immediately, while it takes a while for the answer to make it back to the offerer) it is clear the the addTrack event in some cases could fire as soon as the offerer gets RTP packets with mid indicating to the m-line in question. However, the `RTCTrackEvent` would not be able to contain the `streams[]`data, nor would the track have an Id until the answer has been applied.

It seems to make most sense, in case there is an outstanding offer, to hold off firing `ontrack´ until the answer has arrived? I'm not sure exactly what the spec says though.

But regardless, in this situation there might be a gap in the cloned stream. The question is if the issue is big enough for us to redesign.

**Comment by stefhak** [comment]
I'm closing this Issue now. I don't think we have found enough motivation for changing the spec, @martinthomson please re-open if you disagree.
